ksqldb-common/src/main/java/io/confluent/ksql/schema/ksql/LogicalSchema.java

```diff
@@ -236,48 +269,115 @@ public String toString(final FormatOptions formatOptions) {
     return byNamespace;
   }
 
-  private LogicalSchema rebuild(
-      final boolean withPseudoAndKeyColsInValue,
-      final boolean windowedKey
-  ) {
+  /**
+   * Rebuilds schema with pseudocolumns and key columns included
+   * @param windowedKey indicates if the schema to be rebuilt includes a windowed key
+   * @param pseudoColumnVersion indicates which set of pseudocolumns should be used
+   * @return the LogicalSchema created, with the corresponding pseudo and key columns included
+   */
+  private LogicalSchema rebuildWithPseudoAndKeyColsInValue(
+      final boolean windowedKey, final int pseudoColumnVersion) {
     final Map<Namespace, List<Column>> byNamespace = byNamespace();
 
     final List<Column> key = byNamespace.get(Namespace.KEY);
-    final List<Column> value = byNamespace.get(VALUE);
 
     final ImmutableList.Builder<Column> builder = ImmutableList.builder();
 
-    builder.addAll(key);
+    final List<Column> keyColumns = keyColumns(byNamespace);
 
-    int valueIndex = 0;
-    for (final Column c : value) {
-      if (SystemColumns.isSystemColumn(c.name())) {
-        continue;
-      }
+    final List<Column> nonPseudoAndKeyCols = nonPseudoAndKeyColsAsValueCols(
+        byNamespace, pseudoColumnVersion);
 
-      if (findColumnMatching(withNamespace(Namespace.KEY).and(withName(c.name()))).isPresent()) {
-        continue;
-      }
+    builder.addAll(keyColumns);
+    builder.addAll(nonPseudoAndKeyCols);
+
+    int valueIndex = nonPseudoAndKeyCols.size();
+
+    if (pseudoColumnVersion >= ROWTIME_PSEUDOCOLUMN_VERSION) {
+      builder.add(Column.of(ROWTIME_NAME, ROWTIME_TYPE, VALUE, valueIndex++));
+    }
+
+    if (pseudoColumnVersion >= ROWPARTITION_ROWOFFSET_PSEUDOCOLUMN_VERSION) {
+      builder.add(Column.of(ROWPARTITION_NAME, ROWPARTITION_TYPE, VALUE, valueIndex++));
+      builder.add(Column.of(ROWOFFSET_NAME, ROWOFFSET_TYPE, VALUE, valueIndex++));
+    }
 
+    for (final Column c : key) {
       builder.add(Column.of(c.name(), c.type(), VALUE, valueIndex++));
     }
 
-    if (withPseudoAndKeyColsInValue) {
-      builder.add(Column.of(ROWTIME_NAME, ROWTIME_TYPE, VALUE, valueIndex++));
+    if (windowedKey) {
+      builder.add(
+          Column.of(WINDOWSTART_NAME, WINDOWBOUND_TYPE, VALUE, valueIndex++));
+      builder.add(
+          Column.of(WINDOWEND_NAME, WINDOWBOUND_TYPE, VALUE, valueIndex));
+    }
+
+
+    return new LogicalSchema(builder.build());
+  }
+
+  /**
+   * Rebuilds schema without pseudocolumns or key columns
+   * @return the LogicalSchema created, with the corresponding pseudo and key columns excluded
+   */
+  private LogicalSchema rebuildWithoutPseudoAndKeyColsInValue(final int pseudoColumnVersion) {
+    final ImmutableList.Builder<Column> builder = ImmutableList.builder();
+
+    final List<Column> keyColumns = keyColumns(byNamespace());
+
+    final List<Column> nonPseudoAndKeyCols = nonPseudoAndKeyColsAsValueCols(
+        byNamespace(), pseudoColumnVersion);
```

Comment:
Let's just compute `byNamespace()` once rather than twice, else we're not getting the benefit of the more complicated method signatures for `keyColumns()` and `nonPseudoAndKeyColsAsValueCols()`.